### PR TITLE
add argmax lowering

### PR DESCRIPTION
**Pull request by dan-garvey**
Add argmax lowering from torch to linalg.

Borrows heavily from the tosa lowering. 

Has an issue at the moment, when I run something like:

npcomp-opt [sample_argmax.mlir](https://gist.github.com/dan-garvey/f29271ba686277e3b6aa0592ecfefc54) -torchscript-to-npcomp-backend-pipeline

I get:
`sample_argmax.mlir:6:10: error: 'std.constant' op requires integer result types to be signless
    %1 = torch.aten.argmax %arg1, %int0, %true_0 : !torch.tensor, !torch.int, !torch.bool -> !torch.tensor                                                     
         ^
sample_argmax.mlir:6:10: note: see current operation: %6 = "std.constant"() {value = 0 : si64} : () -> si64`

I struggled for awhile to realize that I had to make a case in refinetypes that changed the output dtype to int, but I'm not sure if I did that correctly. I'm probably missing something really obvious, but I'm not sure how to address this error.

### CODE REVIEW

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 1]
```diff
@@ -761,12 +761,148 @@ static Value createLinalgPayloadCalculationForReduceOp(
...
+     ->convertType(argmaxOp->getResult(0).getType())
+     .cast<RankedTensorType>();
+   //auto resultType = resultTypeGet.cast<ShapedType>();
+   auto inElementType = inputType.getElementType();
+   auto outElementType = resultType.getElementType();
+   int64_t axis;
+   if (!matchPattern(argmaxOp.dim(), m_TorchConstantInt(&axis)))
+      return rewriter.notifyMatchFailure(argmaxOp, "dim must be constant");
+   auto resultMaxType = RankedTensorType::get(resultType.getShape(), inElementType);
+   if (!inputType.hasStaticShape())
```
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:805]: We never write code that is hardcoded to static shapes in npcomp. Please support dynamic shapes. This generally requires inserting `tensor.dim` ops + math to calculate shapes, then feeding dynamic values to `linalg.init_tensor`.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 2]
```diff
@@ -745,6 +749,33 @@ ChangeResult TypeAnalyzer::visitReductionAlongDimIntListOp(
   }
   return getLatticeElement(op->getResult(0)).join(knowledge);
 }
+ChangeResult TypeAnalyzer::visitAtenArgmaxOp(
+    AtenArgmaxOp op, ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
+  auto input = operands[0]->getValue();
+  auto knowledge = ValueKnowledge::getPessimisticValueState(op->getContext());
+  knowledge.dtype = IntegerType::get(op->getContext(), 64, IntegerType::Signed);
+  int64_t dim;
+  bool keepdimBool;
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:758]: nit: call this just keepDim (and feel free to fix up any surrounding code to be consistent)

File: frontends/pytorch/e2e_testing/torchscript/argmax.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([-1, -1], torch.float32, True),
+    ])
+    def forward(self, a):
+        return torch.argmax(a, None)
```
Comment (silvasean) [inline frontends/pytorch/e2e_testing/torchscript/argmax.py:23]: also test the explicit dim case and the keepdim case.

File: include/npcomp/Dialect/Torch/IR/GeneratedAtenOps.td [hunk 4]
```diff
@@ -1295,6 +1295,38 @@ def Torch_AtenArangeStartOp : Torch_Op<"aten.arange.start", [
   let assemblyFormat = "$start `,` $end `,` $dtype `,` $layout `,` $device `,` $pin_memory attr-dict `:` type($start) `,` type($end) `,` type($dtype) `,` type($layout) `,` type($device) `,` type($pin_memory) `->` type($result)";
 }
 
+def Torch_AtenArgmaxOp : Torch_Op<"aten.argmax", [
```
Comment (silvasean) [inline include/npcomp/Dialect/Torch/IR/GeneratedAtenOps.td:1298]: where are the corresponding changes to torch_ods_gen.py

File: include/npcomp/Dialect/Torch/IR/GeneratedAtenOps.td [hunk 5]
```diff
@@ -1295,6 +1295,38 @@ def Torch_AtenArangeStartOp : Torch_Op<"aten.arange.start", [
...
+    TorchOptionalIntType:$dim,
+    Torch_BoolType:$keepdim
+  );
+  let results = (outs
+    AnyTorchTensorType:$result
+  );
+  let assemblyFormat = "$self `,` $dim `,` $keepdim attr-dict `:` type($self) `,` type($dim) `,` type($keepdim) `->` type($result)";
+}
+
+def Torch_AtenArgmaxOutOp : Torch_Op<"aten.argmax.out", [
```
Comment (silvasean) [inline include/npcomp/Dialect/Torch/IR/GeneratedAtenOps.td:1314]: The .out variant won't work with the code we currently have. ReduceOpVariants needs to learn to support it before it is meaningful to add (it should be handled similar to `trailing_` variants, see emit_with_mutating_variants in torch_ods_gen.py and ReduceOpVariants pass)

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 6]
```diff
@@ -745,6 +749,33 @@ ChangeResult TypeAnalyzer::visitReductionAlongDimIntListOp(
   }
   return getLatticeElement(op->getResult(0)).join(knowledge);
 }
+ChangeResult TypeAnalyzer::visitAtenArgmaxOp(
+    AtenArgmaxOp op, ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
+  auto input = operands[0]->getValue();
+  auto knowledge = ValueKnowledge::getPessimisticValueState(op->getContext());
+  knowledge.dtype = IntegerType::get(op->getContext(), 64, IntegerType::Signed);
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:756]: `si64` is correct here :+1:

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 7]
```diff
@@ -761,12 +761,148 @@ static Value createLinalgPayloadCalculationForReduceOp(
...
+         argmaxOp, "unsupported aten.argmax element type");
+
+   auto fillValueMax = rewriter.create<ConstantOp>(loc, fillValueMaxAttr);
+   auto filledTensorMax =
+       rewriter.create<linalg::FillOp>(loc, fillValueMax, initTensorMax)
+           .result();
+
+   // We need to reduce along the arg-max axis, with parallel operations along
+   // the rest.
+   SmallVector<StringRef, 4> iteratorTypes;
```
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:850]: nit: you don't need the explicit `, 4` here. Same for the other SmallVector's please.

**Review by silvasean — CHANGES_REQUESTED** [review]
You are the first to step on the landmine of integer types in TorchToLinalg :) You will need to add a case for it in [setupValueTensorToBuiltinTensorConversion](https://github.com/llvm/mlir-npcomp/blob/6724de7692284f63ea15a1ea86018e93adb06b15/lib/Dialect/TorchConversion/Transforms/BackendTypeConversion.cpp#L37) which transitively calls into [ValueTensorType::toBuiltinTensor](https://github.com/llvm/mlir-npcomp/blob/6724de7692284f63ea15a1ea86018e93adb06b15/lib/Dialect/Torch/IR/TorchTypes.cpp#L273) -- that function needs to be updated to translate the dtypes to builtin integer types. Specifically, float types get passed through, and integer types get converted to a signless integer version of the same bitwidth.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 6 above)
Comment (cathyzhyi) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:756]: Seems the dtype can be other types as well? Should we just use the dtype from the input?

Comment (silvasean) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:756]: argmax returns the indices of the value, not the value itself:
https://pytorch.org/docs/stable/generated/torch.argmax.html

Comment (cathyzhyi) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:756]: ha, I see.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 2 above)
Comment (dan-garvey) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp]: done

File: frontends/pytorch/e2e_testing/torchscript/argmax.py (hunk 3 above)
Comment (dan-garvey) [reply frontends/pytorch/e2e_testing/torchscript/argmax.py]: done

File: include/npcomp/Dialect/Torch/IR/GeneratedAtenOps.td (hunk 5 above)
Comment (dan-garvey) [reply include/npcomp/Dialect/Torch/IR/GeneratedAtenOps.td]: thanks
